compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt

```diff
@@ -1,287 +1,382 @@
 
-    private fun createBridgeHeader(
-        irClass: IrClass,
-        target: IrSimpleFunction,
-        signatureFunction: IrSimpleFunction,
-        isSpecial: Boolean,
-        isSynthetic: Boolean
-    ): IrSimpleFunction {
-        val modality = if (isSpecial && !target.isCollectionStub()) Modality.FINAL else Modality.OPEN
-        val origin = if (isSynthetic) IrDeclarationOrigin.BRIDGE else IrDeclarationOrigin.BRIDGE_SPECIAL
-
-        val visibility = if (signatureFunction.visibility === Visibilities.INTERNAL) Visibilities.PUBLIC else signatureFunction.visibility
-        val descriptor = WrappedSimpleFunctionDescriptor()
-        return IrFunctionImpl(
-            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
-            origin,
-            IrSimpleFunctionSymbolImpl(descriptor),
-            Name.identifier(signatureFunction.getJvmName()),
-            visibility,
-            modality,
-            returnType = signatureFunction.returnType.eraseTypeParameters(),
-            isInline = false,
-            isExternal = false,
-            isTailrec = false,
-            isSuspend = signatureFunction.isSuspend,
-            isExpect = false,
-            isFakeOverride = origin == IrDeclarationOrigin.FAKE_OVERRIDE,
-            isOperator = false
-        ).apply {
-            descriptor.bind(this)
-            parent = irClass
-            copyTypeParametersFrom(target)
-
-            // Have to specify type explicitly to prevent an attempt to remap it.
-            dispatchReceiverParameter = irClass.thisReceiver?.copyTo(this, type = irClass.defaultType)
-            extensionReceiverParameter = signatureFunction.extensionReceiverParameter
-                ?.copyWithTypeErasure(this)
-            signatureFunction.valueParameters.mapIndexed { i, param ->
-                valueParameters.add(i, param.copyWithTypeErasure(this))
-            }
+    private fun IrClass.addBridge(bridge: Bridge, target: IrSimpleFunction): IrSimpleFunction =
+        addFunction {
+            modality = Modality.OPEN
+            origin = IrDeclarationOrigin.BRIDGE
+            // Internal functions can be overridden by non-internal functions, which changes their names since the names of internal
+            // functions are mangled. In order to avoid mangling the name twice we reset the visibility for bridges to internal
+            // functions to public and use the mangled name directly.
+            visibility = bridge.overridden.visibility.takeUnless { it == Visibilities.INTERNAL } ?: Visibilities.PUBLIC
+            name = Name.identifier(bridge.signature.name)
+            returnType = bridge.overridden.returnType.eraseTypeParameters()
+            isSuspend = bridge.overridden.isSuspend
+        }.apply {
+            copyParametersWithErasure(this@addBridge, bridge.overridden)
+            body = context.createIrBuilder(symbol).run { irExprBody(delegatingCall(this@apply, target)) }
+            overriddenSymbols += bridge.overridden.symbol
```

Comment:
I think this should be fine, since the overridden symbol is for a method with the same signature as the bridge we are creating. We should get the correct return type boxing this way, and afaik this is the only reason to keep `overriddenSymbols` after bridge lowering (default arguments are already resolved).

I see two alternatives. We could add all the overridden symbols to every bridge, which might lead to too much return type boxing and could result in missing bridges. Or we could add all the overridden symbols with the same signature to the bridge as overridden symbols (and then filter out redundant overrides). The latter would be cleaner, but I'm not sure whether it is actually needed.

Reply:
It does not influence anything right now, but I still prefer the second option. The reason is, IR is a language, with its own validity conditions (even if they are not written out anywhere). Every time we violate those conditions, it might bite us at some point in the future, so the less we do it, the better.

Reply:
True enough, that has certainly happened more than once. I've fixed the code to keep all relevant overridden symbols.